cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp

```diff
@@ -108,6 +159,95 @@ sycl::event compute_logloss(sycl::queue& q,
     return loss_event;
 }
 
+template <typename Float>
+sycl::event compute_logloss_with_der_sparse(sycl::queue& q,
+                                            sparse_matrix_handle& sp_handler,
+                                            const ndview<std::int32_t, 1>& labels,
+                                            const ndview<Float, 1>& probabilities,
+                                            ndview<Float, 1>& out,
+                                            ndview<Float, 1>& out_derivative,
+                                            bool fit_intercept,
+                                            const event_vector& deps) {
+    ONEDAL_PROFILER_TASK(compute_logloss_with_grad_sparse, q);
+
+    const std::int64_t n = labels.get_dimension(0);
+    const std::int64_t p = out_derivative.get_dimension(0) - (fit_intercept ? 1 : 0);
+
+    ONEDAL_ASSERT(labels.has_data());
+    ONEDAL_ASSERT(probabilities.has_data());
+    ONEDAL_ASSERT(out.has_mutable_data());
+    ONEDAL_ASSERT(out_derivative.has_mutable_data());
+    ONEDAL_ASSERT(out.get_dimension(0) == 1);
+    ONEDAL_ASSERT(probabilities.get_dimension(0) == n);
+
+    auto derivative_object = ndarray<Float, 1>::empty(q, { n }, sycl::usm::alloc::device);
+
+    auto* const der_obj_ptr = derivative_object.get_mutable_data();
+    const auto* const proba_ptr = probabilities.get_data();
+    const auto* const labels_ptr = labels.get_data();
+    auto* const out_ptr = out.get_mutable_data();
+    auto* const out_derivative_ptr = out_derivative.get_mutable_data();
+
+    auto loss_event = q.submit([&](sycl::handler& cgh) {
+        using oneapi::dal::backend::operator+;
+        using sycl::reduction;
+
+        cgh.depends_on(deps);
+        auto sum_reduction_logloss = reduction(out_ptr, sycl::plus<>());
+        const auto wg_size = propose_wg_size(q);
+        const auto range = make_multiple_nd_range_1d(n, wg_size);
+
+        cgh.parallel_for(range, sum_reduction_logloss, [=](sycl::nd_item<1> id, auto& sum_logloss) {
+            auto idx = id.get_group_linear_id() * wg_size + id.get_local_linear_id();
+            if (idx >= std::size_t(n))
+                return;
+            const Float prob = proba_ptr[idx];
+            const float label = labels_ptr[idx];
+            sum_logloss += -label * sycl::log(prob) - (1 - label) * sycl::log(1 - prob);
+            der_obj_ptr[idx] = prob - label;
+        });
+    });
+    //-------
+    loss_event.wait_and_throw();
+    //-------
+    sycl::event derw0_event = sycl::event{};
+    if (fit_intercept) {
+        derw0_event = q.submit([&](sycl::handler& cgh) {
+            using oneapi::dal::backend::operator+;
+            using sycl::reduction;
+
+            cgh.depends_on(deps + loss_event);
```

Comment:
I guess you can delete +loss_event, due to you already wait it

Reply:
Well, I wouldn't do this. In fact, wait and throw for gemv was introduced as a security measure when mkl kernels were not stable and might be removed in future. For this reason it is better to keep the actual dependency graph

Reply:
it is better to align it for the current impl